CrossSectionAnalysis/CrossSectionAnalysis.py

```diff
@@ -1263,32 +1215,54 @@ def updateOutputTable(self, inputCenterline, outputTable):
     logging.info(message)
     slicer.util.showStatusMessage(message, 5000)
 
-  def updatePlot(self, outputPlotSeries, outputTable, name=None):
-
-    # Create plot
-    if name:
-      outputPlotSeries.SetName(name)
-    outputPlotSeries.SetAndObserveTableNodeID(outputTable.GetID())
-    outputPlotSeries.SetXColumnName(DISTANCE_ARRAY_NAME)
-    if self.outputPlotSeriesType == MIS_DIAMETER:
-        outputPlotSeries.SetYColumnName(MIS_DIAMETER_ARRAY_NAME)
-    elif self.outputPlotSeriesType == CE_DIAMETER:
-        outputPlotSeries.SetYColumnName(CE_DIAMETER_ARRAY_NAME)
-    elif self.outputPlotSeriesType == CROSS_SECTION_AREA:
-        outputPlotSeries.SetYColumnName(CROSS_SECTION_AREA_ARRAY_NAME)
-    elif self.outputPlotSeriesType == WALL_CE_DIAMETER:
-        outputPlotSeries.SetYColumnName(WALL_DIAMETER_ARRAY_NAME)
-    elif self.outputPlotSeriesType == WALL_CROSS_SECTION_AREA:
-        outputPlotSeries.SetYColumnName(WALL_CROSS_SECTION_AREA_ARRAY_NAME)
-    elif self.outputPlotSeriesType == DIAMETER_STENOSIS:
-        outputPlotSeries.SetYColumnName(DIAMETER_STENOSIS_ARRAY_NAME)
-    elif self.outputPlotSeriesType == SURFACE_AREA_STENOSIS:
-        outputPlotSeries.SetYColumnName(SURFACE_AREA_STENOSIS_ARRAY_NAME)
-    else:
-      pass
-    outputPlotSeries.SetPlotType(slicer.vtkMRMLPlotSeriesNode.PlotTypeScatter)
-    outputPlotSeries.SetMarkerStyle(slicer.vtkMRMLPlotSeriesNode.MarkerStyleNone)
-    outputPlotSeries.SetColor(0, 0.6, 1.0)
+  def _createOutputPlotSeries(self, name, yColumnName, unit, color = (0.0, 0.6, 1.0)):
+    plotSeriesNode = slicer.mrmlScene.AddNewNodeByClass("vtkMRMLPlotSeriesNode")
+    # It will be weird to get nodes in the python console;
+    # It is nevertheless useful since we can't use a Y axis title.
+    fullName = "{nameOfArray} ({unitOfMeasurement})".format(nameOfArray=yColumnName, unitOfMeasurement=unit)
+    plotSeriesNode.SetName(fullName)
+    plotSeriesNode.SetXColumnName(DISTANCE_ARRAY_NAME)
+    plotSeriesNode.SetYColumnName(yColumnName)
+    plotSeriesNode.SetPlotType(slicer.vtkMRMLPlotSeriesNode.PlotTypeScatter)
+    plotSeriesNode.SetMarkerStyle(slicer.vtkMRMLPlotSeriesNode.MarkerStyleNone)
+    plotSeriesNode.SetColor(color)
+    plotSeriesNode.SetAndObserveTableNodeID(self.outputTableNode.GetID())
+    # Make sure the plot is in the chart
+    if not self.outputPlotChartNode.HasPlotSeriesNodeID(plotSeriesNode.GetID()):
+      self.outputPlotChartNode.AddAndObservePlotSeriesNodeID(plotSeriesNode.GetID())
+    return plotSeriesNode
+
+  def updatePlot(self, outputPlotChart, outputTable, name=None):
+    # Recreate all output plot series nodes.
+    self.outputPlotSeriesNodes.clear()
+    self.outputPlotChartNode.RemoveAllPlotSeriesNodeIDs()
```

Comment:
We should also delete the series nodes from the scene, otherwise we would leave behind a lot of obsolete series nodes each time we update the plot. Of course, we don't want to accidentally delete series that some other modules generated and were added to the same chart, so the module could to set some attribute on the series node that it creates and only delete those series that it generated.

Reply:
The nodes are now removed from scene.

Also:
Using a distinct logic instance for each MRML scripted node. They are tracked and applied when a scripted node is selected.